src/optimizer/cost_and_stats_calculator.cpp

```diff
@@ -43,18 +50,90 @@ std::shared_ptr<TableStats> generateOutputStat(
 
       // TODO: Deal with or add column stats for complex expressions like a+b
       if (expr->GetExpressionType() == ExpressionType::VALUE_TUPLE) {
-        auto column_name =
-            (std::dynamic_pointer_cast<expression::TupleValueExpression>(expr))
-                ->GetColumnName();
+        auto tv_expr =
+            std::dynamic_pointer_cast<expression::TupleValueExpression>(expr);
+        auto column_name = tv_expr->GetColumnName();
+        if (data_table == nullptr) {
+          column_name = tv_expr->GetTableName() + "." + column_name;
+        }
         auto column_stat = input_table_stats->GetColumnStats(column_name);
         if (column_stat != nullptr) {
+          if (data_table != nullptr) {
+            // add table_name for base table
+            column_stat->column_name =
+                data_table->GetName() + "." + column_stat->column_name;
+          }
           output_column_stats.push_back(column_stat);
         };
       }
     }
   }
   auto output_table_stats = std::make_shared<TableStats>(
       input_table_stats->num_rows, output_column_stats);
+
+  // Set sampler
+  output_table_stats->SetTupleSampler(input_table_stats->GetSampler());
+
+  // Add indexes to index map for base table
+  if (data_table != nullptr) {
+    for (oid_t i = 0; i < data_table->GetIndexCount(); i++) {
+      auto index = data_table->GetIndex(i);
+      // If the index is for multiple columns, we don't add
+      if (index->GetMetadata()->GetKeyAttrs().size() > 1) {
+        continue;
+      }
+      auto column_id = index->GetMetadata()->GetKeyAttrs()[0];
+      // If the index column does not exist in the property column, we don't add
+      if (input_table_stats->GetColumnStats(column_id)->column_name.find(".") ==
+          std::string::npos) {
+        continue;
+      }
+      // To use input_table_stats to ensure the column id is correct
+      output_table_stats->AddIndex(
+          input_table_stats->GetColumnStats(column_id)->column_name,
+          data_table->GetIndex(i));
+    }
+  }
+  return output_table_stats;
+}
+
+std::shared_ptr<TableStats> generateOutputStatFromTwoTable(
+    std::shared_ptr<TableStats> &left_table_stats,
+    std::shared_ptr<TableStats> &right_table_stats,
+    const PropertyColumns *columns_prop) {
+  std::vector<std::shared_ptr<ColumnStats>> output_column_stats;
+  if (columns_prop->HasStarExpression()) {
+    for (size_t i = 0; i < left_table_stats->GetColumnCount(); i++) {
+      auto column_stat = left_table_stats->GetColumnStats((oid_t)i);
```

Comment:
If you use oid instead of string here to get columns, then it wouldn't be correct if the table is not base table.

Reply:
what I do here is to add all the column stats to the output stats, so the column_id does not matter. Sorry for my misleading codes here